### PR TITLE
Create dataloader on trainer __init__()

### PR DESCRIPTION
**Pull request by ravi-mosaicml**
#65 made the global rank available in the process start, so it is no longer necessarry to wait until training_start() to create the dataloader. Instead, dataloaders are now initialized in __init__.

This change will help with dataloader profiling, as now the dataloader will be immediately bound to the state.

### CODE REVIEW

File: composer/core/state.py [hunk 1]
```diff
@@ -108,6 +108,10 @@ class State(Serializable):
     # stopping conditions
     max_epochs: int
 
+    # dataloaders
+    train_dataloader: types.DataLoader
+    eval_dataloader: types.DataLoader
```
Comment (ajaysaini725) [inline composer/core/state.py:113]: @anisehsani should be aware of this change since he's working on supporting multiple eval dataloaders

**Review by ajaysaini725 — APPROVED** [review]
This doesn't change the external API at all so LGTM 

I think we should have some discussion around whether to have the `__init__` take any information about datasets at all or have the train + eval datasets be passed directly into `fit()` as well (this is a standard among other ML libraries like sklearn and HuggingFace).

File: composer/core/state.py (hunk 1 above)
Comment (anisehsani) [reply composer/core/state.py:113]: Thanks for the ping. This actually simplifies some thing for me!

Comment (ravi-mosaicml) [reply composer/core/state.py:113]: Would be down to discuss this, though my initial thoughts are to keep fit as a zero-argument function, which can allow `__init__` to be used as a kind of static analysis check.

**Pull request merged**